### ISSUE TITLE
return a different instance from a constructor

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-spec#1344]
It would surely be a popular crowd-pleasing feature to let you write this:

```ceylon
class Point {
    shared Float x;
    shared Float y;
    shared new origin {
        x = 0.0;
        y = 0.0;
    }
    shared new (Float x, Float y) {
        if (x==0.0 && y==0.0) {
            return origin;
        }
        else {
            this.x = x;
            this.y = y;
        }
    }
}
```

On the other hand, from a Java interop POV, it would probably suck.

**Comment by jvasileff** [comment]
I think this would be very nice. A couple notes:

1. May need `final` constructors, to prevent extension by subclasses.
2. This seems to take the *other* interpretation of `return` in #1288 (i.e. return to caller, not skip to end of block)

**Comment by luolong** [comment]
> 2. This seems to take the *other* interpretation of `return` in #1288 (i.e. return to caller, not skip to end of block)

In fact, this is mixed variant, as one branch of the if expression returns an instance of a value constructor and the other one has no return in sight...

**Comment by simonthum** [comment]
In more complex cases, it might be difficult to avoid overallocation (allocating this but returning another instance). I like it but would keep the separation between factory method and constructor, at least as seen from within.

**Comment by jvasileff** [comment]
> keep the separation between factory method and constructor

Yeah, I agree. I think my early enthusiasm on this issue would be better directed at https://github.com/ceylon/ceylon-spec/issues/899.

**Comment by gavinking** [comment]
I note that reaction to this has been overwhelmingly lukewarm.

**Comment by simonthum** [comment]
I very much welcome any way of having easily discoverable factory methods (i.e. *not* rather arbitrarily named functions dumped in the enclosing scope for lack of a nicer place).

However, what worries me is that your example looks like mixing up ctors and factory methods, blurring the line between them.
